### PR TITLE
S20-628, back-end pluggability

### PR DESCRIPTION
**Pull request by taaraora**
1. builtin plugins where removed.
2. plugin install/remove job was added
3. all code but ```handler``` package in ```pkg/api``` package is generated.
4. Plugin protobufs API was actualized
5. plugins registartion/unregistration logic was added to ```pkg/analyze/plugin_controller.go```
6. storage package was improved to support watch API
etc...

### CODE REVIEW

File: pkg/api/handlers/register_plugin_handler.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		storage: storage,
+		log:     logger,
+	}
+}
+
+func (h *registerPluginHandler) Handle(params operations.RegisterPluginParams) middleware.Responder {
+	h.log.Debugf("got request at: %v, request: %+v", time.Now(), params)
+	defer h.log.Debugf("request processing finished at: %v, request: %+v", time.Now(), params)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```
Comment (saheienko) [inline pkg/api/handlers/register_plugin_handler.go]: Will `1s` for timeout be enough here?

File: pkg/api/handlers/unregister_plugin_handler_test.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	req.Header.Set("Content-Type", "application/json")
+	rr := httptest.NewRecorder()
+	h.ServeHTTP(rr, req)
+
+	if status := rr.Code; status != http.StatusBadRequest {
+		t.Fatalf("handler returned wrong status code: got %v want %v, body: %v", status, http.StatusBadRequest, rr.Body.String())
+	}
+}
+
+func TestUnregisterPluginHandler_ReturnNotFound(t *testing.T) {
```
Comment (saheienko) [inline pkg/api/handlers/unregister_plugin_handler_test.go]: Here is a good place for using [TableDrivenTests](https://github.com/golang/go/wiki/TableDrivenTests) approach.

File: cmd/analyzed/app/command.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+	if servingError := server.Serve(); servingError != nil {
+		return errors.Wrap(servingError, "unable to serve HTTP API")
+	}
+
+	return nil
+}
+
+func discoverETCDEndpoint() string {
+	etcdHost, hostExists := os.LookupEnv("ETCD_SERVICE_HOST")
+	etcdPort, portExists := os.LookupEnv("ETCD_SERVICE_PORT")
```
[the remark this answers is not in the document]
Comment (taaraora) [reply cmd/analyzed/app/command.go]: I don't see where it is duplicated. =(

Comment (saheienko) [reply cmd/analyzed/app/command.go]: Yep, everything is ok.

File: pkg/analyze/plugin_controller.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+	return nil
+}
+
+func (pc *PluginController) check(pluginID string, pluginClient *plugin.Client) func() error {
+	return func() error {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*120)
+		checkResponse, err := pluginClient.Check(ctx, &proto.CheckRequest{})
+		if err != nil {
+			cancel()
```
Comment (egopher) [inline pkg/analyze/plugin_controller.go]: This is incorrect usage of cancelling context. If you already got an error from pluginClient it means that the method returned, not timeouted. It is better to defer cancel() instead.

File: cmd/analyzed/app/command.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+	"github.com/supergiant/analyze/pkg/config"
+	"github.com/supergiant/analyze/pkg/kube"
+	"github.com/supergiant/analyze/pkg/logger"
+	"github.com/supergiant/analyze/pkg/models"
+	"github.com/supergiant/analyze/pkg/proxy"
+	"github.com/supergiant/analyze/pkg/scheduler"
+	"github.com/supergiant/analyze/pkg/storage/etcd"
+)
+
+func RunCommand(cmd *cobra.Command, _ []string) error {
```
Comment (egopher) [inline cmd/analyzed/app/command.go]: Why do you need the whole cobra library if there is only one command?

File: cmd/analyze-nodeagent/main.go [hunk 6]
```diff
@@ -42,6 +42,7 @@ func runCommand(cmd *cobra.Command, _ []string) error {
 	}
 
 	logger := logrus.New()
+	logger.SetLevel(logrus.DebugLevel)
```
[the remark this answers is not in the document]
Comment (taaraora) [reply cmd/analyze-nodeagent/main.go]: Fixed.

File: cmd/analyze-registry-job/main.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+		return errors.Wrap(err, "unable to get config flag remove")
+	}
+
+	pluginServiceName, err := discoverPluginServiceName()
+	if err != nil {
+		return errors.Wrap(err, "unable to get plugin service name")
+	}
+
+	logger := logrus.New()
+	logger.SetLevel(logrus.DebugLevel)
```
[the remark this answers is not in the document]
Comment (taaraora) [reply cmd/analyze-registry-job/main.go]: Fixed.

File: cmd/analyze-registry-job/main.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+		return errors.Wrap(err, "unable to create kube client")
+	}
+
+	pluginService, err := kubeClient.GetService(pluginServiceName, pluginLabelSet)
+	if err != nil {
+		return errors.Wrap(err, "failed to find analyze service")
+	}
+
+	var pluginApiPort string
+	for _, port := range pluginService.Spec.Ports {
```
[the remark this answers is not in the document]
Comment (taaraora) [reply cmd/analyze-registry-job/main.go]: That is not a protocol schema. That is just the name of a port in service object. I can name it ```analyze-api``` or use any other string.

File: cmd/analyze-registry-job/main.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+	logger.Debugf("analyze service name %v, service port %v", analyzeServiceName, analyzeApiPort)
+
+	pi.ServiceEndpoint = pluginServiceName + ":" + pluginApiPort
+	pi.ServiceLabels = pluginService.Labels
+	bytes, err := json.Marshal(pi)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal plugin info")
+	}
+
+	pluginsEndpointUri := "http://" + analyzeServiceName + ":" + analyzeApiPort + "/api/v1/plugins"
```
[the remark this answers is not in the document]
Comment (taaraora) [reply cmd/analyze-registry-job/main.go]: Currently, internally I use only http. And if we start to use https we need to add some logic which 
 will discover schema.

File: cmd/analyze-registry-job/main.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+	if !(resp.StatusCode == http.StatusCreated || resp.StatusCode == http.StatusOK) {
+		return errors.Errorf("failed to register plugin, status code %v", resp.StatusCode)
+	}
+
+	resp, err = http.Get(pluginsEndpointUri)
+	if err != nil {
+		return errors.Wrap(err, "unable to check registered plugin")
+	}
+
+	result := make([]*models.Plugin, 0, 0)
```
[the remark this answers is not in the document]
Comment (taaraora) [reply cmd/analyze-registry-job/main.go]: Fixed.

File: pkg/analyze/plugin_controller.go [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+			if err := pc.handleEvent(we); err != nil {
+				pc.logger.Errorf("unable to handle watch event, err: %+v", err)
+			}
+		}
+	}
+}
+
+// TODO: maybe split config and plugin updates?
+func (pc *PluginController) handleEvent(we storage.WatchEvent) error {
+	if we.Type() == storage.Error {
```
[the remark this answers is not in the document]
Comment (taaraora) [reply pkg/analyze/plugin_controller.go]: I didn't get your question. There is no such possibility. If a user does issue ```helm install supergiant/analyze``` multiple times the one will get separate etcd instances for each Analyze instance. If a user is enough experienced to reconfigure Analyze to work with specific etcd, yes multiple Analyze instance will register and invoke the same plugins, but an experienced user will not do such things.

File: pkg/analyze/plugin_controller.go [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+	if err != nil {
+		return errors.Wrap(err, "unable to register proxy for plugin")
+	}
+
+	return nil
+}
+
+func (pc *PluginController) check(pluginID string, pluginClient *plugin.Client) func() error {
+	return func() error {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*120)
```
[the remark this answers is not in the document]
Comment (taaraora) [reply pkg/analyze/plugin_controller.go]: Please check my previous comment - https://github.com/supergiant/analyze/pull/40#discussion_r258529133

File: pkg/api/handlers/plugin_handler.go [hunk 13]
```diff
@@ -13,26 +13,38 @@ import (
...
+func NewPluginHandler(storage storage.Interface, logger logrus.FieldLogger) operations.GetPluginHandler {
+	return &pluginHandler{
 		storage: storage,
 		log:     logger,
 	}
 }
 
-func (h *pluginsHandler) Handle(params operations.GetPluginsParams) middleware.Responder {
+func (h *pluginHandler) Handle(params operations.GetPluginParams) middleware.Responder {
 	h.log.Debugf("got request at: %v, request: %+v", time.Now(), params)
```
[the remark this answers is not in the document]
Comment (taaraora) [reply pkg/api/handlers/plugin_handler.go]: There is TODO for that task.

File: pkg/api/handlers/plugin_handler.go [hunk 14]
```diff
@@ -13,26 +13,38 @@ import (
...
+	defer h.log.Debugf("request processing finished at: %v, request: %+v", time.Now(), params)
+
 	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```
[the remark this answers is not in the document]
Comment (taaraora) [reply pkg/api/handlers/plugin_handler.go]: Please check my previous comment - https://github.com/supergiant/analyze/pull/40#discussion_r258529133

File: pkg/api/handlers/register_plugin_handler.go [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+	if params.Body == nil || "" == strings.TrimSpace(params.Body.ID) {
+		r := operations.NewRegisterPluginDefault(http.StatusBadRequest)
+		message := "plugin id can't be empty"
+		r.Payload = &models.Error{
+			Code:    http.StatusBadRequest,
+			Message: &message,
+		}
+		return r
+	}
+	p := &models.Plugin{
```
[the remark this answers is not in the document]
Comment (taaraora) [reply pkg/api/handlers/register_plugin_handler.go]: Please read a few lines above.

File: pkg/analyze/plugin_controller.go [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+			cancel()
+			return errors.Errorf("unable to marshal check result, pluginClient: %s, returned error: %s", pluginID, err)
+		}
+
+		err = pc.stor.Put(ctx, models.CheckResultPrefix, pluginID, msg(bytes))
+		if err != nil {
+			cancel()
+			return errors.Errorf("unable to store check result, pluginClient: %s, returned error: %s", pluginID, err)
+		}
+		cancel()
```
[the remark this answers is not in the document]
Comment (taaraora) [reply pkg/analyze/plugin_controller.go]: It was one more execution path where cancel was needed to be invoked. I've changed it to defer.

File: cmd/analyzed/app/command.go (hunk 5 above)
Comment (taaraora) [reply cmd/analyzed/app/command.go]: Makes sense, I've removed cobra.

**Pull request merged**